File: MATLAB/step03.m

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Step 3: 1D Diffusion
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

clc; clear; close;

 X = 41;                   % Number of spatial points
 T = 20;                   % Number of time steps
 v = 0.1;                  % Diffusion coefficient
dx = 2 / (X - 1);          % Spatial step size
dt = 0.2 * dx^2 / v;       % Time step size


% Initialise spatial grid and initial condition
u = ones(1, X);     
u(1, round(0.5 / dx):round(1 / dx + 1)) = 2;  
x = linspace(0, 2, X);

for n = 1:T
    un = u;
    for i = 2:X-1
        u(i) = v * dt/dx^2 * (un(i+1) - 2*un(i) + un(i-1)) + un(i);
    end

    % Plot
    hFig = figure(1);
    set(hFig, 'Position', [360 278 560 249])
    plot(x, u, 'LineWidth', 1);
    axis([0, 2, 0.5, 2.5]);
    title(sprintf('Time Step: %d', n),'fontsize',14);
    set(0,'defaultTextInterpreter','latex');
    xlabel('$x$','fontsize',14);
    ylabel('$u$','fontsize',14);
end

File: C++/step03.cpp
#include "matplotlibcpp.h"
#include <cmath>
#include <vector>
#include <algorithm>
#include <iostream>

namespace plt = matplotlibcpp;

////////////////////////////////////////////////////////////
// Step 3: 1D Diffusion
////////////////////////////////////////////////////////////

int main() {
    // Simulation parameters
    const int X = 41;                    // Number of spatial points
    const int T = 20;                    // Number of time steps

    const double  v = 0.1;               // Diffusion coefficient
    const double dx = 2.0 / (X - 1);     // Spatial step size
    const double dt = 0.2 * (dx * dx) / v;    // Time step size

    // Initialize spatial grid and initial condition
    std::vector<double> x(X), u(X), un(X);

    for (int i = 0; i < X; i++) {
        x[i] = (2.0 * i) / (X - 1);
        u[i] = 1 + (x[i] >= 0.5 && x[i] <= 1);
    }

    plt::ion();
    plt::Plot plot;

    // Time-stepping loop
    for (int n = 0; n < T; n++) {
        un = u;

        for (int i = 1; i < X-1; i++) {
            u[i] = v * dt/(dx * dx) * (un[i+1] - 2*un[i] + un[i-1]) + un[i];
        }

        plot.update(x, u);
        plt::xlim(0, 2);
        plt::ylim(0.5, 2.5);
        plt::pause(0.1);
    }

    plt::show();
    return 0;
}
